docs/how-to-use-monty/tutorials/implementing-actions.md

```diff
@@ -0,0 +1,161 @@
+```python
+class Jump(Action):
+
+    def __init__(self, agent_id: str, how_high: float) -> None:
+        super().__init__(agent_id=agent_id)
+        self.how_high = how_high
+```
+
+If at any point this action will be read from a file, then be sure to update the `ActionJSONDecoder` as well:
+
+```python
+class ActionJSONDecoder(JSONDecoder):
+
+    # ...
+
+    def object_hook(self, obj: dict[str, Any]) -> Any:
+        # ...
+        elif action = Jump.action_name():
+            return Jump(
+                agent_id=obj["agent_id"]
+                how_high=obj["how_high"]
+            )
+        # ...
+```
+
+With the above in place, the **Motor System** can now create the **Action** as needed.
+
+# Sampling Actions
+
+For the **Motor System** to be able to sample the **Action** with the help of a sampler, you will need to include a `sample` method and declare an action-specific sampler protocol:
+
+```python
+class JumpActionSampler(Protocol):
+    def sample_jump(self, agent_id: str) -> Jump: ...
+
+class Jump(Action):
+
+    @staticmethod
+    def sample(agent_id: str, sampler: JumpActionSampler) -> Jump:
+        return sampler.sample_jump(agent_id)
+
+    def __init__(self, agent_id: str, how_high: float) -> None:
+        super().__init__(agent_id=agent_id)
+        self.how_high = how_high
+```
+
+This prepares the **Action** to be used in a sampler.
+
+The sampler itself will subclass `ActionSampler` and implement the requisite protocol:
+
+```python
+class MyConstantSampler(ActionSampler):
+
+    def __init__(
+        self,
+        actions: list[type[Action]] | None = None,
+        height: float = 1.8,
+        rng: Generator | None = None
+    ) -> None:
+        super().__init__(actions=actions, rng=rng)
+        self.height = 1.8
+
+    def sample_jump(self, agent_id: str) -> Jump:
+        return Jump(agent_id=agent_id, how_high=self.height)
+```
+
+The sampler then will be used along the lines of:
+
+```python
+sampler = MyConstantSampler(actions=[Jump])
+action = sampler.sample("agent_0")
+```
+
+For already existing samplers, see `src/tbp/monty/frameworks/actions/action_samplers.py`.
+
+# Actuating Actions
+
+For an **Action** to take effect within an **Environment**, it needs to be actuated. Include the `act` method and declare an action-specific actuator protocol:
+
+```python
+class JumpActuator(Protocol):
+    def actuate_jump(self, action: Jump) -> None: ...
+
+class Jump(Action):
+
+    def __init__(self, agent_id: str, how_high: float) -> None:
+        super().__init__(agent_id=agent_id)
+        self.how_high = how_high
+
+    def act(self, actuator: JumpActuator) -> None:
+        actuator.actuate_jump(self)
+```
+
+The actuator itself will be specific to the **Environment** and the simulator, robot, or otherwise. However, in general, it will implement the actuator protocol:
+
+```python
+class MyActuator:
+
+    def actuate_jump(self, action: Jump) -> None: ...
+        # custom code to make Jump happen
+```
```

Comment:
I'm not sure this code block adds anything that isn't already in the previous code block? Why not add the comment into the previous code block or make it a bit more concrete with some example code in there?

Reply:
This wasn't meant to be a protocol definition. I updated it to

```python
class MyActuator:

    def actuate_jump(self, action: Jump) -> None:
        # custom code to make Jump happen
        pass
```

This being a general guide, implementing actuation could be anything. For an example implementation, readers are directed to the habitat actuator at the bottom of the section.

Reply:
That's a good point, I'll move this over.

Reply:
https://github.com/thousandbrainsproject/tbp.monty/pull/334